pkg/virtual/syncer/controllers/apireconciler-controller.go

```diff
@@ -0,0 +1,201 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	apiresourcev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apiresource/v1alpha1"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	apiresourceinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apiresource/v1alpha1"
+	apiresourcelistersv1alpha1 "github.com/kcp-dev/kcp/pkg/client/listers/apiresource/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/virtual/syncer"
+)
+
+// NewAPIReconciler returns a new controller which reconciles APIResourceImport resources
+// and delegates the corresponding WorkloadClusterAPI management to the given WorkloadClusterAPIManager.
+func NewAPIReconciler(
+	apiManager syncer.WorkloadClusterAPIManager,
+	kcpClusterClient kcpclient.ClusterInterface,
+	apiResourceImportInformer apiresourceinformer.APIResourceImportInformer,
+	negotiatedAPIResourceInformer apiresourceinformer.NegotiatedAPIResourceInformer,
+) (*APIReconciler, error) {
+	name := "kcp-virtual-syncer-api-reconciler"
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), name)
+
+	c := &APIReconciler{
+		name:                        name,
+		apiWatcher:                  apiManager,
+		kcpClusterClient:            kcpClusterClient,
+		apiresourceImportLister:     apiResourceImportInformer.Lister(),
+		negotiatedAPIResourceLister: negotiatedAPIResourceInformer.Lister(),
+		queue:                       queue,
+		workloadClusterAPIs:         make(map[string]syncer.WorkloadClusterAPI),
+	}
+
+	apiResourceImportInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueue(obj)
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			c.enqueue(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueue(obj)
+		},
+	})
+
+	return c, nil
+}
+
+// APIReconciler is a controller that reconciles APIResourceImport resources
+// and delegates the corresponding WorkloadClusterAPI management to the given WorkloadClusterAPIManager.
+type APIReconciler struct {
+	name                        string
+	apiWatcher                  syncer.WorkloadClusterAPIManager
+	kcpClusterClient            kcpclient.ClusterInterface
+	apiresourceImportLister     apiresourcelistersv1alpha1.APIResourceImportLister
+	negotiatedAPIResourceLister apiresourcelistersv1alpha1.NegotiatedAPIResourceLister
+
+	queue workqueue.RateLimitingInterface
+
+	apiMutex            sync.RWMutex
+	workloadClusterAPIs map[string]syncer.WorkloadClusterAPI
```

Comment:
I don't see the value of this state. We have another copy in the WorkloadClusterAPIManager.

Reply:
resolved via controller clean up and merge of controller and installedapis.